frontend/class-handle-404.php

```diff
@@ -0,0 +1,50 @@
+<?php
+/**
+ * WPSEO plugin file.
+ *
+ * @package WPSEO\Frontend
+ */
+
+/**
+ * Class WPSEO_Handle_404
+ *
+ * Handles intercepting requests.
+ *
+ * @since 1.8
+ */
+class WPSEO_Handle_404 implements WPSEO_WordPress_Integration {
+	/**
+	 * Registers all hooks to WordPress
+	 *
+	 * @return void
+	 */
+	public function register_hooks() {
+		add_filter( 'pre_handle_404', array( $this, 'filter_feed_404s' ), 10, 2 );
+	}
+
+	/**
+	 * If there are no posts in a feed, make it 404 instead of sending an empty RSS feed.
+	 *
+	 * @param boolean  $handled  Whether we've handled the request.
+	 * @param WP_Query $wp_query The Query object.
+	 *
+	 * @return bool
+	 */
+	public function filter_feed_404s( $handled, $wp_query ) {
+		global $wp_query;
+		if ( is_feed() ) {
+			if ( $wp_query->found_posts === 0 ) {
```

Comment:
@jdevalk I've created better checking. Just checking found_posts isn't enough. Also, it could send 404 errors if archive is "empty", but it should return 200.
```php
if ( $wp_query->get_queried_object() && ( $wp_query->is_singular() || $wp_query->is_archive() ) && ! $wp_query->is_date() )
       || ( $wp_query->is_date() && empty( $wp_query->posts ) )
) {
```

Reply:
Did you do a pull? :)

Reply:
I've prepared the code :smiley:, but I thought that's something which should be [fixed in WP core](https://core.trac.wordpress.org/ticket/30210). Anyway, I could create PR during the weekend.